File: 1_Algorithmic_Toolbox/week2_algorithmic_warmup/3_greatest_common_divisor/GCD.java
import java.util.Scanner;

public class GCD {
    static final boolean FULL_TEST = false;

    private static int gcd_naive(int a, int b) {
        int current_gcd = 1;
        for (int d = 2; d <= a && d <= b; ++d) {
            if (a % d == 0 && b % d == 0) {
                if (d > current_gcd) {
                    current_gcd = d;
                }
            }
        }

        return current_gcd;
    }

    private static int gcd(int a, int b) {
        if (a > b) {
            return gcd(b, a);
        }
        if (a == 0)
            return b;
        return gcd(b % a, a);
    }

    public static void main(String args[]) {
        if (FULL_TEST) {
            int MAX_TEST_VAL = 20;
            for (int i = 1; i <= MAX_TEST_VAL; ++i) {
                for (int j = 1; j < MAX_TEST_VAL; j++)
                    System.out.println(
                            String.format("(%d,%d) gcd == %d, gcd_naive == %d", i, j, gcd(i, j), gcd_naive(i, j)));
            }
        } else {
            Scanner scanner = new Scanner(System.in);
            int a = scanner.nextInt();
            int b = scanner.nextInt();

            System.out.println(gcd(a, b));
        }
    }
}

File: 1_Algorithmic_Toolbox/week2_algorithmic_warmup/3_greatest_common_divisor/gcd.py
# Uses python3
import sys

DEBUG = False

def gcd_naive(a, b):
    current_gcd = 1
    for d in range(2, min(a, b) + 1):
        if a % d == 0 and b % d == 0:
            if d > current_gcd:
                current_gcd = d

    return current_gcd

# Reimplement gcd_fast using Euclid's algorithm
def gcd_fast(a, b):
    """
    Calculates the greatest common divisor of two numbers.

    Parameters:
    a (int): The first number.
    b (int): The second number.

    Returns:
    int: The greatest common divisor of the two numbers.

    Examples:
    >>> gcd_fast(18, 35)
    1
    >>> gcd_fast(28851538, 1183019)
    17657
    """
    # Euclid's algorithm
    # Divide the larger number by the smaller number
    # If the remainder is 0, then the smaller number is the GCD
    # If the remainder is not 0, then the smaller number is the new larger number
    # and the remainder is the new smaller number
    if a < b:
        a, b = b, a
    while b != 0:
        a, b = b, a % b
    return a

# Write test cases
def test_cases():
    assert gcd_fast(18, 35) == 1
    assert gcd_fast(28851538, 1183019) == 17657
    print('All test cases passed.')

if __name__ == "__main__":
    if DEBUG:
        test_cases()
    else:
        input = sys.stdin.read()
        a, b = map(int, input.split())
        print(gcd_fast(a, b))
